nuxeo-features/nuxeo-platform-oauth/src/main/java/org/nuxeo/ecm/platform/oauth2/clients/OAuth2Client.java

```diff
@@ -18,23 +18,64 @@
  */
 package org.nuxeo.ecm.platform.oauth2.clients;
 
+import static java.util.Objects.requireNonNull;
+import static javax.servlet.http.HttpServletResponse.SC_BAD_REQUEST;
 import static org.nuxeo.ecm.platform.oauth2.clients.OAuth2ClientService.OAUTH2CLIENT_SCHEMA;
 
 import java.util.Arrays;
 import java.util.Collections;
+import java.util.HashMap;
 import java.util.List;
+import java.util.Map;
 import java.util.Optional;
 import java.util.regex.Pattern;
 
 import org.apache.commons.lang3.StringUtils;
 import org.nuxeo.ecm.core.api.DocumentModel;
+import org.nuxeo.ecm.core.api.NuxeoException;
+import org.nuxeo.ecm.directory.BaseSession;
 
 /**
  * @author <a href="mailto:ak@nuxeo.com">Arnaud Kervern</a>
  * @since 5.9.2
  */
 public class OAuth2Client {
 
+    /**
+     * @since 11.1
+     */
+    public static final String NAME_FIELD = "name";
+
+    /**
+     * @since 11.1
+     */
+    public static final String ID_FIELD = "clientId";
+
+    /**
+     * @since 11.1
+     */
+    public static final String SECRET_FIELD = "clientSecret";
+
+    /**
+     * @since 11.1
+     */
+    public static final String REDIRECT_URI_FIELD = "redirectURIs";
+
+    /**
+     * @since 11.1
+     */
+    public static final String AUTO_GRANT_FIELD = "autoGrant";
+
+    /**
+     * @since 11.1
+     */
+    public static final String ENABLED_FIELD = "enabled";
```

Comment:
I wanted to do it, but I was thinking of doing it on another refactoring commit. This why i create the constants to avoid duplication of these fields. Do you think i can do it on the same commit ? i means the main one or a new one is required ?

Reply:
in addition, i have a suggestion about doc.getProperty, can we add (not in this same jira) a new method on DocumentModel that will be generic
```
<T> Optional<T> getProperty(String schema, , String name, Class<T> clazz) {
     return Optional.ofNullable(this.getProperty(schema, name)).map(clazz::cast);
}
```
i think it will more clear then do the cast manually like ;
 (String) doc.getPropertyValue(OAUTH2CLIENT_SCHEMA + ":name");
or 
(String) doc.getPropertyValue(schema,  ":name");

Reply:
You can do the refactoring in the same commit, especially in this case because it could be less performant without the refactoring.

About the API on `DocumentModel`, returning an `Optional` could be cool, but I find `getProperty(String, String, Class)` signature a bit verbose. We could (probably) change the current API without breaking compatibility to:
```
<T extends Serializable> T getPropertyValue(String, String)
```
but we can't move to `Optional` with same parameters (backward compatibility). A solution could be to have:
```
<T extends Serializable> Optional<T> getOptPropertyValue(String, String)
```
Nevertheless, not sure if we want to do one of them. Maybe @efge has some clue about this API improvement?

Reply:
yes i agree ;)

Reply:
I don't like passing class arguments just so that we can hide a generic cast that will crash in the same way anyway if the expected type is wrong. So for instance I prefer (and don't find more verbose, and it's actually shorter):

    (String) doc.getPropertyValue(myprop);

to:

    doc.getPropertyValue(myprop, String.class);

At least for these kinds of API where the class isn't used for anything more than the cast.

Reply:
BTW I'm often drowned in PR comments/notifications so for important questions please ping me in Slack or by email.